File: scripts/generate_dep.pl
my $core_color = "goldenrod";
my $entrypoint_color = "steelblue1";
my $algorithmic_color = "darkturquoise";
my $syntactic_color = "darkorange";
my $semantic_color = "forestgreen";
my $soundness_color = "goldenrod1";
my $completeness_color = "deeppink3";
my $frontend_color = "darkslategray4";
my $extraction_color = "darkslategray3";
my $others_color = "lightpink";
print "digraph Mcltt {\n";
print "  graph [cluster=true,fontsize=28,label=\"Mcltt\",labeljust=l,labelloc=t,penwidth=5,size=15,splines=true,tooltip=\"\"];\n";
print "  node [fontsize=18,shape=note,style=filled,URL=\"https://beluga-lang.github.io/McLTT/\\N.html\"];\n";
print "  subgraph Algorithmic { label=\"Algorithmic\"; fontcolor=$algorithmic_color; color=$algorithmic_color }\n";
print "  subgraph Core { label=\"Core\"; fontcolor=$core_color; color=$core_color }\n";
print "  subgraph Core { subgraph Core_basic { style=invis; subgraph Core_Syntactic { style=\"\"; label=\"Syntactic\"; fontcolor=$syntactic_color; color=$syntactic_color } } }\n";
print "  subgraph Core { subgraph Core_basic { subgraph Core_Semantic { style=\"\"; label=\"Semantic\"; fontcolor=$semantic_color; color=$semantic_color } } }\n";
print "  subgraph Core { subgraph Core_proof { style=invis; subgraph Core_completeness { subgraph Core_Completeness { style=\"\"; label=\"Completeness\"; fontcolor=$completeness_color; color=$completeness_color } } } }\n";
print "  subgraph Core { subgraph Core_proof { subgraph Core_soundness { subgraph Core_Soundness { style=\"\"; label=\"Soundness\"; fontcolor=$soundness_color; color=$soundness_color } } } }\n";
print "  subgraph Extraction { label=\"Extraction\"; fontcolor=$extraction_color; color=$extraction_color }\n";
print "  subgraph Frontend { label=\"Frontend\"; fontcolor=$frontend_color; color=$frontend_color }\n";
while (<>) {
    if (m/([^\s]*)\.vo.*:(.*)/) {
        my $goal = `realpath $1`;
        $dests_rel = $2 ;
        ($path,$src) = ($goal =~ s/\//\./rg =~ m/.*\.theories\.(.*\.)?([^.]*)\n$/);
        my $goalmodule = "Mcltt.$path$src";
        for my $dest_rel (split(" ", $dests_rel)) {
            my $dest = `realpath $dest_rel`;
            $dest =~ s/\//\./g ;
            print "  \"Mcltt.$1\" -> \"$goalmodule\" [arrowtail=inv,dir=both,penwidth=3,tooltip=\"$goalmodule requires Mcltt.$1\"];\n" if ($dest =~ m/.*\.theories\.(.*)\.vo\n/);
        }
        my $nodeformat = "\"$goalmodule\"[label=\"%s\",tooltip=\"$goalmodule\",fillcolor=%s]";
        if ($path =~ m/Algorithmic\.(.*)/) {
            printf "subgraph Algorithmic { $nodeformat }\n", "$1$src", "$algorithmic_color";
        } elsif ($path =~ m/Syntactic\.(.*)/) {
            printf "subgraph Core { subgraph Core_basic { subgraph Core_Syntactic { $nodeformat } } }\n", "$1$src", "$syntactic_color"
        # } elsif ($goalmodule =~ m/Mcltt\.Core\.Semantic\.Consequences$/) {
        #     printf "subgraph Core { subgraph Core_basic { subgraph Core_Semantic { subgraph Core_Semantic_consequence { cluster=false; rank=sink; $nodeformat} } } }\n", "$1$src", "$semantic_color"
        } elsif ($path =~ m/Semantic\.(.*)/) {
            printf "subgraph Core { subgraph Core_basic { subgraph Core_Semantic { $nodeformat } } }\n", "$1$src", "$semantic_color"
        } elsif ($path =~ m/Soundness\.(.*)/) {
            printf "subgraph Core { subgraph Core_proof { subgraph Core_soundness { subgraph Core_Soundness { $nodeformat } } } }\n", "$1$src", "$soundness_color"
        } elsif ($src =~ m/Soundness$/) {
            # printf "subgraph Core { subgraph Core_proof { subgraph Core_soundness { subgraph soundness { cluster=false; rank=sink; $nodeformat } } } }\n", "$src", "$soundness_color"
            printf "subgraph Core { subgraph Core_proof { subgraph Core_soundness { $nodeformat } } }\n", "$src", "$soundness_color"
        # } elsif ($path =~ m/Completeness\.(Consequences\..*)/) {
        #     printf "subgraph Core { subgraph Core_proof { subgraph Core_completeness { subgraph Core_Completeness { subgraph Core_Completeness_consequences { cluster=false; rank=sink; $nodeformat } } } } }\n", "$1$src", "$completeness_color"
        } elsif ($path =~ m/Completeness\.(.*)/) {
            printf "subgraph Core { subgraph Core_proof { subgraph Core_completeness { subgraph Core_Completeness { $nodeformat } } } }\n", "$1$src", "$completeness_color"
        } elsif ($src =~ m/Completeness$/) {
            # printf "subgraph Core { subgraph Core_proof { subgraph Core_completeness { subgraph completeness { cluster=false; rank=sink; $nodeformat } } } }\n", "$src", "$completeness_color"
            printf "subgraph Core { subgraph Core_proof { subgraph Core_completeness { $nodeformat } } }\n", "$src", "$completeness_color"
        } elsif ($path =~ m/Frontend\.(.*)/) {
            printf "subgraph Frontend { $nodeformat }\n", "$1$src", "$frontend_color"
        } elsif ($src =~ m/Entrypoint$/) {
            printf "subgraph entrypoint { cluster=false; rank=sink; $nodeformat }\n", "$src", "$entrypoint_color"
        } elsif ($path =~ m/Extraction\.(.*)/) {
            printf "subgraph Extraction { $nodeformat }\n", "$1$src", "$extraction_color"
        } elsif ($path =~ m/Core\.(.*)$/) {
            printf "subgraph Core { subgraph Core_basic { $nodeformat } }\n", "$1$src", "$core_color"
        } else {
            printf "subgraph others { cluster=false; rank=source; $nodeformat }\n", "$path$src", "$others_color"
        }
    }
}
print "}\n";

File: scripts/generate_dep.py
import itertools, os, re, sys, textwrap
from pathlib import Path
from typing import Iterable

PROJECT_ROOT = Path(__file__).resolve().parents[1]
THEORIES_ROOT = PROJECT_ROOT / "theories"
ROOT_MODULE_NAME = "Mctt"
COLORS = {
    "Algorithmic": "darkturquoise",
    "Completeness": "deeppink3",
    "Core": "goldenrod",
    "Entrypoint": "steelblue1",
    "Syntactic": "darkorange",
    "Semantic": "forestgreen",
    "Soundness": "goldenrod1",
    "Frontend": "darkslategray4",
    "Extraction": "darkslategray3",
    "others": "lightpink",
}

GRAPH_SET: set[str] = set()
RANKED_GRAPH_SET: set[str] = set()
DOC_BASE = os.environ["DOC_BASE"] if "DOC_BASE" in os.environ else ""

def under_subgraph(name: str, body: str) -> str:
    return f"""subgraph "{name}" {{ {body} }}"""

def subgraph_decl(*, name: str, label: str = "", tooltip: str = "", color: str, body: str = "") -> str:
    GRAPH_SET.add(name)
    return under_subgraph(name, f"""graph[color={color},fontcolor={color},label="{label}",tooltip="{tooltip}"]; node[fillcolor={color}]; {body}""")

def default_subgraph_decl(name: str, body: str = "") -> str:
    return subgraph_decl(name=name, label=name, tooltip=f"{ROOT_MODULE_NAME}.{name}", color=COLORS[name], body=body)

def core_subgraph_decl(name: str, body: str = "") -> str:
    return under_subgraph("Core", subgraph_decl(name=f"Core/{name}", label=name, tooltip=f"{ROOT_MODULE_NAME}.Core.{name}", color=COLORS[name], body=body))

DEPLINE_PATTERN = re.compile(r"(\S*\.vo).*:(.*)")
DEPSOURCE_PATTERN = re.compile(r"\S*\.vo")

def module_parts_of_path(path: str) -> tuple[str, ...]:
    return Path(path).resolve().relative_to(THEORIES_ROOT).with_suffix("").parts

def module_name_of_module_parts(parts: Iterable[str]) -> str:
    return f"""{ROOT_MODULE_NAME}.{".".join(parts)}"""

def module_name_of_path(path: str) -> str:
    return module_name_of_module_parts(module_parts_of_path(path))

def node_of_path(path: str) -> str:
    parts = module_parts_of_path(path)
    dedupe_parts = list(dict.fromkeys(parts))
    color = next((COLORS[part] for part in reversed(dedupe_parts) if part in COLORS),COLORS["others"])
    subgraph_names = list(itertools.accumulate(parts[:-1], func="{0}/{1}".format))
    module_name = module_name_of_module_parts(parts)
    node_label = parts[-1]

    # Handle special case for two main theorems
    # so that printed graph looks better
    if module_name == "Mctt.Core.Completeness" or module_name == "Mctt.Core.Soundness":
        result = under_subgraph(f"Core/{node_label}", f""""{module_name}"[label="{node_label}",tooltip="{module_name}",color={color},fillcolor=white];""")
    elif module_name == "Mctt.LibTactics":
        result = f"""{{ graph[cluster=false,rank=min]; "{module_name}"[label="{node_label}",tooltip="{module_name}",fillcolor={color}]; }}"""
    elif module_name == "Mctt.Core.Semantic.Consequences":
        result = f"""{{ cluster=false; rank=max; "{module_name}"[label="{node_label}",tooltip="{module_name}",fillcolor={color}]; }}"""
    else:
        result = f""""{module_name}"[label="{node_label}",tooltip="{module_name}",fillcolor={color}];"""

    for subgraph_name in reversed(subgraph_names):
        if subgraph_name in GRAPH_SET:
            if subgraph_name in RANKED_GRAPH_SET:
                result = under_subgraph(f"{subgraph_name}/rank", result)
            result = under_subgraph(subgraph_name, result)
        else:
            GRAPH_SET.add(subgraph_name)
            RANKED_GRAPH_SET.add(subgraph_name)
            result = under_subgraph(f"{subgraph_name}/rank", f"graph[cluster=false,rank=same]; {result}")
            result = under_subgraph(subgraph_name, f"""graph[cluster=true,label={parts[subgraph_names.index(subgraph_name)]},tooltip="{module_name_of_module_parts(subgraph_name.split("/"))}"]; {result}""")

    return result

def data_of_depline(depline: str) -> str:
    if result := DEPLINE_PATTERN.fullmatch(depline.strip()):
        sink_module = module_name_of_path(result.group(1))
        source_paths = result.group(2).split(" ")
        source_modules = (module_name_of_path(source_path) for source_path in source_paths if DEPSOURCE_PATTERN.match(source_path))
        edges = "\n".join(f""""{source_module}" -> "{sink_module}";""" for source_module in source_modules)
        return f"""{edges}{node_of_path(result.group(1))}"""
    else:
        raise ValueError(f"Broken Dependency: \"{depline}\"")

def gen_graph() -> str:
    newline = "\n"
    return textwrap.dedent(f"""
      digraph Mctt {{
        graph [center=true,class="depgraph",cluster=true,fontname="Open Sans",fontsize=28,label="Mctt",labeljust=l,labelloc=t,penwidth=2,size=15,splines=true,tooltip=""];
        node [fontsize=18,shape=note,style=filled,URL="https://beluga-lang.github.io/McTT/{DOC_BASE}/\\N.html"];
        {default_subgraph_decl("Algorithmic")}
        {default_subgraph_decl("Core")}
        {core_subgraph_decl("Completeness")}
        {core_subgraph_decl("Semantic")}
        {core_subgraph_decl("Soundness")}
        {core_subgraph_decl("Syntactic")}
        {default_subgraph_decl("Extraction")}
        {default_subgraph_decl("Frontend")}
        {textwrap.indent(newline.join(data_of_depline(depline) for depline in sys.stdin), "        ").lstrip()}
      }}""")

print(gen_graph())
